book/src/rfcs/foreign_field_mul.md

```diff
@@ -0,0 +1,1273 @@
+We can deal with this non-zero issue by computing carry witness values.
+
+## Computing carry witnesses values $v_0$ and $v_1$
+
+Instead of constraining $h_0$ and $h_1$ to zero, there must be satisfying witness $v_0$ and $v_1$ such that the following constraints hold.
+> 5. There exists $v_0$ such that $h_0 = v_0 \cdot 2^{2\ell}$
+> 6. There exists $v_1$ such that $h_1 = v_1 \cdot 2^{\ell} - v_0$
+
+Here $v_0$ is the last two bits of $h_0$'s $2\ell + 2$ bits, i.e., the result of adding the highest bit of $p_0$ and any possible carry bit from the operation of $h_0$.  Similarly, $v_1$ corresponds to the highest $\ell + 3$ bits of $h_1$.  It looks like this
+
+```text
+0             L             2L            3L            4L
+|-------------|-------------|-------------|-------------|-------------|
+                            :
+|--------------h0-----------:--| 2L + 2
+                            : ↖v0
+                            :-------------h1-------------| 2L + 3
+                            :              \____________/
+                            :                  v1➚
+```
+
+
+Remember we only need to prove the first $3\ell$ bits of $p - r$ are zero, since everything is $\mod 2^t$ and  $t = 3\ell$.  It may not be clear how this approach proves the $3\ell$ bits are indeed zero because within $h_0$ and $h_1$ there are bits that are nonzero.  The key observation is that these bits are too high for $\mod 2^t$.
+
+By making the argument with $v_0$ and $v_1$ we are proving that $h_0$ is something where the $2\ell$ least significant bits are all zeros and that $h_1 + v_0$ is something where the $\ell$ are also zeros.  Any nonzero bits after $3\ell$ do not matter, since everything is $\mod 2^t$.
+
+All that remains is to range check $v_0$ and $v_1$
+> 7. Range check $v_0 \in [0, 2^2)$
+> 8. Range check $v_1 =\in [0, 2^{\ell + 3})$
+
+## Subtractions
+
+When unconstrained, computing $u_0 = p_0 + 2^{\ell} \cdot p_{10} - (r_0 + 2^{\ell} \cdot r_1)$ could require borrowing.  Fortunately, we have the constraint that the $2\ell$ least significant bits of $u_0$ are `0` (i.e. $u_0 = 2^{2\ell} \cdot v_0$), which means borrowing cannot happen.
+
+Borrowing is prevented because when the the $2\ell$ least significant bits of the result are `0` it is not possible for the minuend to be less than the subtrahend.  We can prove this by contradiction.
+
+Let
+
+* $x = p_0 + 2^{\ell} \cdot p_{10}$
+* $y = r_0 + 2^{\ell} \cdot r_1$
+* the $2\ell$ least significant bits of $x - y$ be `0`
+
+Suppose that borrowing occurs, that is, that $x < y$.  Recall that the length of $x$ is $2\ell + 2$ bits.  Therefore, since $x < y$ the top two bits of $x$ must be zero and so we have
+
+$$
+x - y = x_{2\ell} - y,
+$$
+
+where $x_{2\ell}$ denotes the $2\ell$ least significant bits of $x$.
+
+Recall also that the length of $y$ is $2\ell$ bits.  So the result of this subtraction is $2\ell$ bits.   Since the $2\ell$ least significant bits of the subtraction are `0` this means that
```

Comment:
Perhaps you want to point out that we know this because limbs of the result are constrained to be a certain range.

Reply:
Will do!